zebrad/src/components/tokio.rs

```diff
@@ -23,3 +25,87 @@ impl TokioComponent {
         })
     }
 }
+
+/// Zebrad's graceful shutdown function, blocks until one of the supported
+/// shutdown signals is received.
+async fn shutdown() {
+    imp::shutdown().await;
+}
+
+/// Extension trait to centralize entry point for runnable subcommands that
+/// depend on tokio
+pub(crate) trait RuntimeRun {
+    fn run(&mut self, fut: impl Future<Output = Result<(), Report>>);
+}
+
+impl RuntimeRun for Runtime {
+    fn run(&mut self, fut: impl Future<Output = Result<(), Report>>) {
+        let result = self.block_on(async move {
+            tokio::select! {
+                result = fut => result,
+                _ = shutdown() => Ok(()),
+            }
+        });
+
+        match result {
+            Ok(()) => {}
+            Err(e) => {
+                eprintln!("Error: {:?}", e);
+                app_writer().shutdown(Shutdown::Forced);
+            }
+        }
+    }
+}
+
+#[cfg(unix)]
+mod imp {
+    use tokio::signal::unix::{signal, SignalKind};
+    use tracing::info;
+
+    pub(super) async fn shutdown() {
+        tokio::select! {
+            // SIGINT  - To allow Ctrl-c to emulate SIGTERM while developing.
+            () = sig(SignalKind::interrupt(), "SIGINT") => {}
+            // SIGTERM - Kubernetes sends this to start a graceful shutdown.
+            () = sig(SignalKind::terminate(), "SIGTERM") => {}
+        };
+    }
+
+    async fn sig(kind: SignalKind, name: &'static str) {
+        // Create a Future that completes the first
+        // time the process receives 'sig'.
+        signal(kind)
+            .expect("Failed to register signal handler")
+            .recv()
+            .await;
+
+        info!(
+            // use target to remove 'imp' from output
+            target: "zebrad::signal",
+            "received {}, starting shutdown",
+            name,
+        );
+    }
+}
+
+#[cfg(not(unix))]
+mod imp {
+    use futures::prelude::*;
+    use tracing::info;
+
+    pub(super) async fn shutdown() {
+        // On Windows, we don't have all the signals, but Windows also
+        // isn't our expected deployment target. This implementation allows
+        // developers on Windows to simulate proxy graceful shutdown
+        // by pressing Ctrl-C.
```

Comment:
This is specific to linkerd's deployment scenarios.

```suggestion
        //  Wait for Ctrl-C in Windows terminals.
        // (Zebra doesn't support NT Service control messages. Use a service wrapper for long-running instances.)
```